engine/api/workflow/process.go

```diff
@@ -0,0 +1,233 @@
+			}
+			for j := range node.Triggers {
+				t := &node.Triggers[j]
+				//TODO Check conditions
+
+				//Keep the subnumber of the previous node in the graph
+				log.Debug("processWorkflowRun> starting from trigger %#v", t)
+				if err := processWorkflowNodeRun(db, w, &t.WorkflowDestNode, int(nodeRun.SubNumber), []int64{nodeRun.ID}, nil, nil); err != nil {
+					sdk.WrapError(err, "processWorkflowRun> Unable to process node ID=%d", t.WorkflowDestNode.ID)
+				}
+			}
+		}
+	}
+
+	//Checks the joins
+	for i := range w.Workflow.Joins {
+		j := &w.Workflow.Joins[i]
+		sources := map[int64]*sdk.WorkflowNodeRun{}
+
+		//we have to check noderun for every sources
+		for _, id := range j.SourceNodeIDs {
+			sources[id] = nil
+			for x := range w.WorkflowNodeRuns {
+				nodeRun := &w.WorkflowNodeRuns[x]
+				if nodeRun.WorkflowNodeID == id {
+					//We found the source in the list of the noderuns
+					sources[id] = nodeRun
+				}
+			}
+		}
+
+		//now checks if all sources have been completed
+		var ok = true
+		nodeRunIDs := []int64{}
+		for _, nodeRun := range sources {
+			if nodeRun == nil {
+				//One of the sources have not been started
+				ok = false
+				break
+			}
+			if nodeRun.Status != string(sdk.StatusSuccess) && nodeRun.Status != string(sdk.StatusFail) {
+				//One of the sources have not been completed
+				ok = false
+				break
+			}
+			nodeRunIDs = append(nodeRunIDs, nodeRun.ID)
+		}
+
+		if ok {
+			//Keep a ref to the sources
+			nodeRun := sources[j.SourceNodeIDs[0]]
+			if nodeRun == nil {
+				return fmt.Errorf("This should not append...")
+			}
+
+			//All the sources are completed
+			//Checks the triggers
+			for x := range j.Triggers {
+				t := &j.Triggers[x]
+				//TODO Check conditions
+
+				//Keep the subnumber of the previous node in the graph
+				if err := processWorkflowNodeRun(db, w, &t.WorkflowDestNode, int(nodeRun.SubNumber), nodeRunIDs, nil, nil); err != nil {
+					sdk.WrapError(err, "processWorkflowRun> Unable to process node ID=%d", t.WorkflowDestNode.ID)
+				}
+			}
+		}
+
+	}
+
+	if err := updateWorkflowRun(db, w); err != nil {
+		sdk.WrapError(err, "processWorkflowRun>")
+	}
+
+	return nil
+}
+
+func processWorkflowNodeRun(db gorp.SqlExecutor, w *sdk.WorkflowRun, n *sdk.WorkflowNode, subnumber int, sourceNodeRuns []int64, h *sdk.WorkflowNodeRunHookEvent, m *sdk.WorkflowNodeRunManual) error {
+	t0 := time.Now()
+	log.Debug("processWorkflowNodeRun> Begin [#%d.%d]%s.%d", w.Number, subnumber, w.Workflow.Name, n.ID)
+	defer func() {
+		log.Debug("processWorkflowNodeRun> End [#%d.%d]%s.%d  - %.3fs", w.Number, subnumber, w.Workflow.Name, n.ID, time.Since(t0).Seconds())
+	}()
+
+	run := &sdk.WorkflowNodeRun{
+		LastModified:   time.Now(),
+		Start:          time.Now(),
+		Number:         w.Number,
+		SubNumber:      int64(subnumber),
+		WorkflowRunID:  w.ID,
+		WorkflowNodeID: n.ID,
+		Status:         string(sdk.StatusWaiting),
+		Stages:         n.Pipeline.Stages,
+	}
+
+	for i := range run.Stages {
```

Comment:
finally i let workflow.execute to set status of each stage sequentally.
it means that not triggered stages have empty status